Near-Bugfix: Optimization: Minimize the number of times it is checked that no money...

jtimon opened the pull request:
...is created by individual transactions to 2 places (but call only once in each):

- ConnectBlock ( before calculated fees per txs twice )
- AcceptToMemoryPoolWorker ( before called CheckTxInputs 4 times and calculated
   fees per tx one extra time )

Also call tx.GetValueOut() only once per call of CheckTxInputs (instead of 2)

For more motivation: 

~~https://github.com/bitcoin/bitcoin/blob/master/src/main.cpp#L1493~~
https://github.com/jtimon/bitcoin/compare/0.13-consensus-inputs...jtimon:0.13-consensus-inputs-comments

EDIT: partially replaces #6445

Near-Bugfix as pointed out in https://github.com/bitcoin/bitcoin/pull/8498#discussion_r124346132

jtimon committed:
Optimization: Minimize the number of times it is checked that no money is created

by individual transactions to 2 places (but call only once in each):

- ConnectBlock ( before calculated fees per txs twice )
- AcceptToMemoryPoolWorker ( before called CheckTxInputs 4 times and calculated
   fees per tx one extra time )

Also call tx.GetValueOut() only once per call of CheckTxInputs (instead of 2)

dcousens commented on src/main.cpp:
```diff
@@ -1923,16 +1923,14 @@ int GetSpendHeight(const CCoinsViewCache& inputs)
+    // are the actual inputs available?
+    if (!inputs.HaveInputs(tx))
+        return state.DoS(100, false, REJECT_INVALID, "bad-txns-inputs-missingorspent", false,
+                         strprintf("%s: inputs missing/spent", __func__));
```
nit: weird spacing here?

jtimon replied:
Oh, that's how the editor indented by default. That's actually what we have in some other places, but having a glance searching for strprintf in main.cpp I see we're not being consistent with indentation of function calls broken into several lines

Example from the opening parenthesis position:
https://github.com/bitcoin/bitcoin/blob/master/src/main.cpp#L1354
Example 4 spaces from the beginning:
https://github.com/bitcoin/bitcoin/blob/master/src/main.cpp#L1110
Example freestyle: 
https://github.com/bitcoin/bitcoin/blob/master/src/main.cpp#L1436

I'm happy to change it for something else, this maybe?
```
        return state.DoS(100, false, REJECT_INVALID, "bad-txns-inputs-missingorspent", false,
            strprintf("%s: inputs missing/spent", __func__));
```

jtimon commented:
Needed rebase. Besides, the previous version contained a bug.

jtimon commented:
Needed rebase after renaming main.o, see #9260 (although needed, the rebase was clean in this case)

sipa commented on src/validation.cpp:
```diff
@@ -646,18 +646,20 @@ bool AcceptToMemoryPoolWorker(CTxMemPool& pool, CValidationState& state, const C
 
-        // are the actual inputs available?
+        // This redundant check doesn't trigger the DoS code on purpose; if it did, it would make it easier
+        // for an attacker to attempt to split the network (Consensus::CheckTxInputs also checks this).
```
I think you can change the `view.HaveInputs(tx)` below into a `Consensus::CheckTxInputs` call, doing some of the checks slightly earlier.

jtimon replied:
Perhaps the new comment could be clearer, but the ```view.HaveInputs(tx)``` check below sets ```state.DoS(0, ...``` instead of ```state.DoS(100, ...``` like ```Consensus::CheckTxInputs```  would do.
Feel more than free to rephrase it in a way that would have been more clear to you.

sipa reviewed (commented):
Concept ACK

dcousens commented on src/validation.cpp:
```diff
@@ -1318,16 +1318,14 @@ int GetSpendHeight(const CCoinsViewCache& inputs)
+    // are the actual inputs available?
+    if (!inputs.HaveInputs(tx))
+        return state.DoS(100, false, REJECT_INVALID, "bad-txns-inputs-missingorspent", false,
+                         strprintf("%s: inputs missing/spent", __func__));
```
Indentation is borked

dcousens commented on src/validation.cpp:
```diff
@@ -1318,16 +1318,14 @@ int GetSpendHeight(const CCoinsViewCache& inputs)
 
-namespace Consensus {
-bool CheckTxInputs(const CTransaction& tx, CValidationState& state, const CCoinsViewCache& inputs, int nSpendHeight)
+bool Consensus::CheckTxInputs(const CTransaction& tx, CValidationState& state, const CCoinsViewCache& inputs, int nSpendHeight, CAmount& nFees)
```
Rather than forcing a bogus `nFees` (or unused) everywhere,  just wrap this function with a unused `nFees` stack variable.
Cleaner,  and less diff.

jtimon replied:
The only place where this makes sense is in txmempool.cpp. There you can do a much nicer wrapper there. The diff won't be smaller in total but there will be more lines in red (which people tend to like). I considered this before and I rembember @sipa said it would be fine to just remove the check in txmempool.cpp but I can't find the comment and he changed his mind about it. I will rewrite this for you to see and I'm happy to squash it.

dcousens reviewed (approved):
utACK,  but see comment RE suggested change

jtimon committed:
Introduce CheckInputsAndUpdateCoins static wrapper in txmempool.cpp

jtimon commented:
Added a commit that may make @dcousens happier or not, thanks for making me remember.

ryanofsky commented on src/validation.cpp:
```diff
@@ -1820,6 +1814,8 @@ static bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockInd
     std::vector<PrecomputedTransactionData> txdata;
     txdata.reserve(block.vtx.size()); // Required so that pointers to individual PrecomputedTransactionData don't get invalidated
+    const int nHeight = pindex->pprev == NULL ? 0 : pindex->pprev->nHeight + 1;
+    const int64_t nSpendHeight = block.vtx.size() > 1 ? GetSpendHeight(view) : nHeight;
```
In commit "Optimization: Minimize the number of times it is checked"

Can you add a comment explaining this line? Why do transactions in a block with vtx.size() of >= 2 have a different spend height than transactions in a block with vtx.size() == 1? Is nSpendHeight even used in a block with vtx.size() == 1?

jtimon replied:
If the block only has the coinbase tx before GetSpendHeight(view)  would have not been called at all and starting doing so seems unnecessarily costly, see https://github.com/bitcoin/bitcoin/blob/master/src/validation.cpp#L1306
I was actually hoping that someone suggested to just remove this local variable directly and just use nHeight directly for the coinbase maturity check, assuming that is safe. I think so but would like others to confirm before trying it.

ryanofsky reviewed (commented):
utACK 1590c19d3b4fa5d406cf2f91772960e11d0d2cb8

PR was initially confusing to me, but here are my notes on what it does:

- Changes `CheckTxInputs` to return DoS level `100` and code `REJECT_INVALID` instead of `0` and `0`.
- Adds `CheckTxInputs` `nFees` in/out argument
- Removes `CheckTxInputs` call from `CheckInputs`. Avoids changing behavior by having the two callers of `CheckInputs` (`AcceptToMemoryPoolWorker` and `ConnectBlock`) call `CheckTxInputs` themselves.
- Changes `AcceptToMemoryPoolWorker` and `ConnectBlock` to use fees returned by `CheckTxInputs` instead of computing fees themselves.
- Updates calls to `CheckTxInputs` in `CTxMemPool::check`. No impact there because mempool code ignores returned validation state and fees.

jtimon commented:
Needed rebase, hopefully fixed all nits.

ryanofsky commented:
Tests for this PR seem to be failing currently, maybe due to the SpendHeight change?

Only changes since my previous review were rebasing post-#8329 and the nFees and nSpendHeight changes commented on above.

jtimon commented:
Yes, the problem was in that change but not on stop using GetSpendHeight, just added +1 to the height by mistake. Updated.

jtimon commented:
Needed rebase after #10195, it also needs more review now (AcceptToMemoryPoolWorker may return different DoS pnctuation [100 instead of 0] now, happy to fix if that's a problem).

ryanofsky commented:
Some discussion of this in IRC https://botbot.me/freenode/bitcoin-core-dev/msg/87002756/

ryanofsky commented on src/validation.cpp:
```diff
@@ -1622,9 +1619,8 @@ static bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockInd
-            if (!view.HaveInputs(tx))
-                return state.DoS(100, error("ConnectBlock(): inputs missing/spent"),
-                                 REJECT_INVALID, "bad-txns-inputs-missingorspent");
+            if (!Consensus::CheckTxInputs(tx, state, view, pindex->nHeight, nFees))
```
In commit "Optimization: Minimize the number of times"

If I understand correctly, it would be more correct to pass `SpendHeight(view)` instead of `pindex->nHeight` as the nSpendHeight argument, but in this case it doesn't matter because the value is only used for coinbase transactions, and this isn't a coinbase transaction? Or maybe there are other reasons why `pindex->nHeight` is ok to pass?

Either way, I think this is confusing, and that there should be a comment here explaining the `pindex->nHeight` value. Or maybe you could delete the `nSpendHeight` argument and have `CheckTxInputs` call `GetSpendHeight(inputs)` when needed. This seems like it would be a simplification and I don't think there should be a performance cost because `CheckTxInputs` seems to be skipped for coinbase transactions in all cases except for one call that invokes `GetSpendHeight` anyway.

jtimon replied:
I don't think it would be more correct (but I'm glad to be corrected) and performance would sufffer.
I think pindex->nHeight because that's actually what we want to validate here. In acceptToMemPool and CTxMemPool::check you don't know the height. Here, nSpendHeight should always be lower or equal to pindex->nHeight.

I'm happy to add a comment but not sure what it should be. Peharps the comment should be in GetSpendHeight instead of ConnectBlock().

ryanofsky replied:
> I'm happy to add a comment but not sure what it should be.

Comment could just say why it is better to pass pindex->nHeight as nSpendHeight here instead of GetSpendHeight(view). If the two values are always equal, or have some other relation, comment could just say what that relationship is. If this is too in the weeds, definitely feel free to skip this. I'm just suggesting what I think would make the code clearer for me.

Alternately, I don't know what you think of my suggestion to eliminate the `nSpendHeight` argument and just call `GetSpendHeight` in `CheckTxInputs`. It does look to me like this would be logically equivalent and not effect performance, though maybe there's another reason not to do it.

jtimon replied:
Part of the optimization here is not calling GetSpendHeight for every tx in the block, and your suggestion would eliminate that part of the optimization.
For the comment, what about "We know the height, so we don't need to GetSpendHeight"?
it if more people agree.
I'm really not sure that comment would add that much value, but I don't mind adding

sdaftuar replied:
I agree with @jtimon that it makes more sense for ConnectBlock to pass in pindex->nHeight, rather than get the spend height from the view, even though those are the same.  We could add an assert that the view's best block has the same hash as pindex->pprev if that makes the code clearer -- but we have bigger problems than just coinbase maturity if the view is out of sync with our chain!

sdaftuar replied:
style nit: curly braces for one-line if

ryanofsky reviewed (commented):
utACK 4d9e9a5b7cb3e65bcf96ea5090ab6abaeba27640. Only changes since last review were removing a stray comment and fixing the spendheight failures.

ryanofsky reviewed (commented):
Not sure if this needs more review. This PR has two utACKs from dcousens and me, and a concept ack from sipa.

sdaftuar commented on src/consensus/tx_verify.cpp:
```diff
@@ -201,15 +201,14 @@ bool CheckTransaction(const CTransaction& tx, CValidationState &state, bool fChe
-        if (!inputs.HaveInputs(tx))
-            return state.Invalid(false, 0, "", "Inputs unavailable");
+    // are the actual inputs available?
+    if (!inputs.HaveInputs(tx))
```
style nit: braces for one-line if

sdaftuar reviewed (changes requested):
Concept ACK.  I think this will slightly speed up validation time; I'm doing a benchmark to see if this is observable.  Left one request and some style nits below.

jtimon commented on src/consensus/tx_verify.h:
```diff
@@ -24,7 +26,7 @@ namespace Consensus {
  * Preconditions: tx.IsCoinBase() is false.
  */
-bool CheckTxInputs(const CTransaction& tx, CValidationState& state, const CCoinsViewCache& inputs, int nSpendHeight);
+bool CheckTxInputs(const CTransaction& tx, CValidationState& state, const CCoinsViewCache& inputs, int nSpendHeight, CAmount& nFees);
```
If the fee is returned, then we can't also return a bool, and we would need to duplicate the check in ConnectBlock, AcceptToMemeoryPool and CTxMemPool::check. And at that point the additional complexity wouldn't be worth it IMO.

We could also pass a fees input/output parameter that doesn't accumulate (although you can already enjoy that functionality by simply passing 0 to the variable). In that case, the duplication would only need to duplicateMoneyrange check in ConnectBlock, but I don't really see it as an improvement over this.
 
Agreed, accumulated_fees would be more readable with only 2 more lines needing to change

sipa replied:
@jtimon You can 'return` the fee in a tuple, or as a pass-by-reference value. So for example `CheckTxInputs` would have `CAmount& return_fee` field, and `return_fee` would just be *assigned* that transaction's fee. The caller can then do the `accumulated_fee += return_fee` logic. This involves no duplication, and still makes `CheckTxInputs` purely operate on data for a single transaction.

jtimon replied:
If the concern is that the addition with other fees is not done within the function, as documented the caller can set the in/out argument to zero before calling. I could also always return only the value for the single transaction without accumulating anything (although I don't see the gain, it's just more code), at that point the argument would be out only, not in/out.

From there, to return the output argument as a tuple or pair I think it's just making things uglier and unnecessarily complicated.

instead of:
```
CAmount tx_fee = 0; // You will initialize the variable even if the function starts setting it to zero
if (!CheckTxInputs(..., tx_fee))
   return false;
accumulated_fee += tx_fee;
```

You would have something like:

```
struct CheckTxInputsReturn
{
   bool fValid;
   CAmount tx_fee;
}

CheckTxInputsReturn ret = CheckTxInputs(...);
if (!ret.fValid)
   return false;
accumulated_fee += ret.tx_fee;
```

I think that's incredibly ugly. Specially since I don't see the problem with how it is currently used when you don't want to accumulate any fees:

```
CAmount tx_fee = 0;
if (!CheckTxInputs(..., tx_fee))
   return false;
```

Please, I invite you to write your suggestions on top of this and see for yourselves that they are not an improvement but actually the contrary.
If people agree that those IMO much uglier solutions are better, I will squash your suggestions, no problem.

sipa replied:
I was just clarifying that @sdaftuar's suggestion does not imply duplication of logic. I personally think that it's a bit cleaner to do the summing of fees outside of the validation of a single function, but it's an unimportant nit for sure.

jtimon replied:
Oh, sorry, the duplication comment was a mistake. Before I was checking the moneyRange for the accumulated fees and not just for the single tx fees, and I thought that check would need to be duplicated outside, but that was a change in functionality that I shouldn't be doing (and btw was missed in review).
Pedantically adding that check would be a softfork, but it should be impossible that the MoneyRange check on the accumulated fees fails anyway.

I can not accumulate anything if you guys prefer that, that's a simple change (even though I still can't understand why would you prefer that), reading your previous comment again better, you were also considering the pass by reference (what is done now), not necessarily the tuple (which is what I would be against in principle because of its ugliness).

jtimon committed:
Proper indentation for CheckTxInputs and other minor fixes

jtimon commented:
Hopefully fixed @sdaftuar 's nits.
Also add a third commit to properly indent CheckTxInputs and other minor style fixes like braces for one line ifs while at it.

jtimon commented:
Added a couple of commits to be squashed. I wouldn't personally squash the latter but it seems @sdaftuar and @sipa would like it more this way and I don't care. Before squashing I will also move the indentation commit to the first thing.

TheBlueMatt commented on src/validation.cpp:
```diff
@@ -1635,9 +1633,11 @@ static bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockInd
+            if (!Consensus::CheckTxInputs(tx, state, view, pindex->nHeight, tx_fees)) {
+                return error("%s: Consensus::CheckTxInputs: %s, %s", __func__, tx.GetHash().ToString(), FormatStateMessage(state));
+            }
+            nFees += tx_fees;
```
While you're at it, can you please re-add the MoneyRange checks here that Gavin removed years ago (and maybe update the PR title to indicate that you're fixing Gavin's near-bug while also decreasing the number of CCoinsView map lookups).

TheBlueMatt replied:
(way back in 8d7849b6db5f54dc32fe4f8c6c7283068473cd21, for those following along at home)

TheBlueMatt replied:
Specifically, commit 8d7849b6db5f54dc32fe4f8c6c7283068473cd21 broke the check at https://github.com/bitcoin/bitcoin/commit/8d7849b6db#diff-7ec3c68a81efff79b6ca22ac1f1eabbaR1110 and made it useless, which very narrowly didn't cause a major consensus failure and re-introduction of the original bitcoin-printing overflow bug.

TheBlueMatt reviewed (commented):
Totally agree with @sipa and @sdaftuar, thanks for making it a non-accumulator. Generally looks good, obviously needs squash.

jtimon committed:
Near-Bugfix: Reestablish consensus check removed in 8d7849b

in 8d7849b6db5f54dc32fe4f8c6c7283068473cd21

This can potentially prevent an overflow that could at least in theory
allow the creation of money.

jtimon commented:
Squashed the non-accumulator commit.
Hopefully fixed all @TheBlueMatt 's nits, good catch.
Also s/nTxFee/tx_fees/ to comply with the new style.

TheBlueMatt commented on src/validation.cpp:
```diff
@@ -1635,9 +1633,15 @@ static bool ConnectBlock(const CBlock& block, CValidationState& state, CBlockInd
+            nFees += tx_fees;
+            if (!MoneyRange(nFees)) {
+                 state.DoS(100, false, REJECT_INVALID, "bad-txns-accumulated-fee-outofrange");
+                 return error("%s: %s, %s", __func__, tx.GetHash().ToString(), FormatStateMessage(state));
```
Why not use the normal formatting here of return state.DoS(100, error(), ...)?

jtimon replied:
No reason, will change

TheBlueMatt reviewed (commented):
utACK 3161a7c9ef6e19696a79776b38e3e5e4b6a67940

jtimon commented:
Fixed latest nit.

jimpo commented on src/consensus/tx_verify.cpp:
```diff
@@ -205,46 +205,49 @@ bool CheckTransaction(const CTransaction& tx, CValidationState &state, bool fChe
-            return state.DoS(100, false, REJECT_INVALID, "bad-txns-fee-outofrange");
+    // Tally transaction fees
+    tx_fees = nValueIn - value_out;
+    if (tx_fees < 0) {
```
Isn't this redundant given the check on line 239?

jtimon replied:
Yes, I think so, unless I'm missing something about some weird overflow.

jtimon commented:
Hopefully fixed @jimpo 's nit. Didn't squashed yet just in case, but it seems pretty obviously correct in removing a redundancy that cannot be tested and thus is not being tested.

promag commented on src/consensus/tx_verify.h:
```diff
@@ -22,9 +24,10 @@ namespace Consensus {
 /**
  * Check whether all inputs of this transaction are valid (no double spends and amounts)
  * This does not modify the UTXO set. This does not check scripts and sigs.
+ * @param[out] tx_fees this serves to get the fees of the tx as output. 
```
` * @params[out] tx_fees  Set to the transaction fee if successful`.

promag commented on src/consensus/tx_verify.cpp:
```diff
@@ -205,46 +205,46 @@ bool CheckTransaction(const CTransaction& tx, CValidationState &state, bool fChe
+    }
+
+    // Tally transaction fees
+    tx_fees = nValueIn - value_out;
```
Why was/is plural?

jtimon replied:
no good reason, let's make it singular

promag commented on src/txmempool.cpp:
```diff
@@ -611,6 +611,15 @@ void CTxMemPool::clear()
     _clear();
 }
 
+static void CheckInputsAndUpdateCoins(const CTransaction& tx, CCoinsViewCache& mempoolDuplicate, const int64_t nSpendHeight)
```
Use snake_case on new code.

promag commented on src/validation.cpp:
```diff
@@ -531,6 +528,12 @@ static bool AcceptToMemoryPoolWorker(const CChainParams& chainparams, CTxMemPool
+
+        } // end LOCK(pool.cs)
+
+        CAmount nFees = 0;
```
`tx_fees` or just `fees`?

jtimon replied:
That would be more extra disruption only for style. We can do that at any time. Including potentially more rebase work fixing conflicts with other things touching AcceptToMemoryPoolWorker. Can we leave that for later? Note we're not creating a new variable but just moving its declaration here.

promag reviewed (commented):
Some nits.

danra commented:
@jtimon The "More motivation" link https://github.com/bitcoin/bitcoin/blob/master/src/main.cpp#L1493 is broken

jtimon commented:
Fixed @promag 's nits (except one to avoid further disruption, in this case only for style).
Fixed @danra 's nit by crossing the broken list in the OP (too lazy to look after some concept and ut acks, sorry. I believe there's enough motivation without the broken link, which I don't remember what was about right now).

promag commented on src/consensus/tx_verify.cpp:
```diff
@@ -207,44 +207,45 @@ bool CheckTransaction(const CTransaction& tx, CValidationState &state, bool fChe
+
+    CAmount nValueIn = 0;
+    CAmount nFees = 0;
+    for (unsigned int i = 0; i < tx.vin.size(); i++) {
```
`++i`.

promag commented on src/consensus/tx_verify.cpp:
```diff
@@ -205,15 +205,17 @@ bool CheckTransaction(const CTransaction& tx, CValidationState &state, bool fChe
-    // for an attacker to attempt to split the network.
-    if (!inputs.HaveInputs(tx))
-        return state.Invalid(false, 0, "", "Inputs unavailable");
+    txfee = 0; // Initialize output value
```
Why? Update only if successfully?

jtimon replied:
And if not set to to 0? I think this was discussed already...

promag commented on src/consensus/tx_verify.cpp:
```diff
@@ -233,18 +235,15 @@ bool Consensus::CheckTxInputs(const CTransaction& tx, CValidationState& state, c
-    }
-    nFees += nTxFee;
-    if (!MoneyRange(nFees)) {
+    txfee = nValueIn - value_out;
```
Update before `return true;`.

promag commented on src/consensus/tx_verify.cpp:
```diff
@@ -233,18 +235,15 @@ bool Consensus::CheckTxInputs(const CTransaction& tx, CValidationState& state, c
-    nFees += nTxFee;
-    if (!MoneyRange(nFees)) {
+    txfee = nValueIn - value_out;
+    if (!MoneyRange(txfee)) {
```
```cpp
if (!MoneyRange(nValueIn - value_out)) {
```

jtimon replied:
Mhmm, right, that's what we're saying in the doxygen docs now...ok, I guess.

jtimon commented:
Fixed @promag 's last round of nits, thanks again.

danra commented on src/consensus/tx_verify.cpp:
```diff
@@ -205,46 +205,46 @@ bool CheckTransaction(const CTransaction& tx, CValidationState &state, bool fChe
+    }
+
+    CAmount nValueIn = 0;
+    for (unsigned int i = 0; i < tx.vin.size(); ++i) {
```
Suggest using const auto& to iterate over tx.vin. (This is already the case in other functions in this file)

jtimon replied:
This line wasn't touched besides indentation until @promag asked s/i++/++i/
Not sure it's worth it to make that change here, to be honest...

danra replied:
@jtimon IMHO it would fit nicely in the minor changes commit

danra commented on src/consensus/tx_verify.cpp:
```diff
@@ -205,46 +205,46 @@ bool CheckTransaction(const CTransaction& tx, CValidationState &state, bool fChe
+    CAmount nValueIn = 0;
+    for (unsigned int i = 0; i < tx.vin.size(); ++i) {
+        const COutPoint &prevout = tx.vin[i].prevout;
+        const Coin& coin = inputs.AccessCoin(prevout);
```
Suggest using const auto& for these

TheBlueMatt replied:
Please dont add needless auto& unless you have to. If you're saving 5 chars of typing it just makes things less readable.

danra replied:
@TheBlueMatt I disagree, and AFAIK so do most of the C++ gurus. The (character) typing isn't the issue here, rather code flexibility. See https://softwareengineering.stackexchange.com/questions/180216/does-auto-make-c-code-harder-to-understand for an example of a discussion on the topic.

TheBlueMatt replied:
auto, here, AFAIU, is going to be slightly more likely to force you to change the line if you change AccessCoin's return type (at least because its a const-ref type, so conversion should only be an inheritance thing). In consensus code, making sure you're forced to change all the places something is used if you change a type strikes me as a very, very good idea. If you want to take Herb's stance on it, I'd also be ok with his recommended auto coin = const Coin&{inputs.AccessCoin(prevout)}; to force it, though I find that somewhat unreadable (which is bias towards what I'm used to, I admit).

danra replied:
Consensus code seems hard enough to change as it is without adding extra hurdles.

> In consensus code, making sure you're forced to change all the places something is used if
> you change a type strikes me as a very, very good idea.

By that logic, instead of
`const COutPoint &prevout = tx.vin[i].prevout;`

it would be better to write
```
const CTxIn& txin = tx.vin[i];
const COutPoint &prevout = tx.vin[i].prevout;
```
Since in the first version, `tx.vin[i]` is used without specifying its type :)

jtimon replied:
Let's please leave this for another PR

danra replied:
OK, deleting this comment.

danra commented on src/consensus/tx_verify.cpp:
```diff
@@ -205,46 +205,46 @@ bool CheckTransaction(const CTransaction& tx, CValidationState &state, bool fChe
-                return state.DoS(100, false, REJECT_INVALID, "bad-txns-inputvalues-outofrange");
+    // are the actual inputs available?
+    if (!inputs.HaveInputs(tx)) {
+        return state.DoS(100, false, REJECT_INVALID, "bad-txns-inputs-missingorspent", false,
```
The current code has an explicit comment about not triggering the DoS code on purpose, did you remove it (and invoke the DoS code) intentionally?

jtimon replied:
The comment was moved to the separated HaveInputs call in AcceptToMemoryPoolWorker.
Perhaps lost in rebase, I'll look at it.
Perhaps it's fine since we now have: https://github.com/bitcoin/bitcoin/pull/8498/commits/c93ceb2dbde1c2474ee04946fb4594b668613dbe#diff-24efdb00bfbe56b140fb006b562cc70bR514

danra replied:
It's not just the comment, as you can see in the diff previously the code used `state.Invalid` (and stated it does so on purpose), and now both the comment is gone and the code uses `state.DoS`

I don't know to say on my own whether that change is valid, but it definitely requires explanation due to the fact that previously there was a specific comment suggesting using `state.DoS` there is bad.

TheBlueMatt replied:
Oops, yea, cant DoS for ATMP, but should keep the DoS in ConnectBlock (I assume thats why they were split to begin with).

danra replied:
@TheBlueMatt 
> It should be OK, the DoS from ConnectBlock is important
Which DoS? The one previously returned from `CheckTxInputs` in case of missing inputs? If so, IIUC it wouldn't trigger after this change, there would just be an error returned, since it replaced the previous DoS in `ConnectBlock` https://github.com/bitcoin/bitcoin/commit/c93ceb2dbde1c2474ee04946fb4594b668613dbe#diff-24efdb00bfbe56b140fb006b562cc70bR1638

TheBlueMatt replied:
DoS values persist through the state object (and should only be set once through the entire call tree). Thus, when ConnectBlock calls CheckTxInputs, which sets DoS, ConnectBlock *should* return only error() and not call DoS again, but the DoS points will still be returned through the state object.

danra replied:
@TheBlueMatt Ah, I see. Thanks!

danra commented on src/consensus/tx_verify.h:
```diff
@@ -8,6 +8,8 @@
 #include <stdint.h>
 #include <vector>
 
+#include "amount.h"
```
Local includes should come before system includes

jtimon commented:
Fixed 1 commit by @danra

TheBlueMatt commented on src/consensus/tx_verify.cpp:
```diff
@@ -205,46 +205,46 @@ bool CheckTransaction(const CTransaction& tx, CValidationState &state, bool fChe
+    CAmount nValueIn = 0;
+    for (unsigned int i = 0; i < tx.vin.size(); ++i) {
+        const COutPoint &prevout = tx.vin[i].prevout;
+        const Coin& coin = inputs.AccessCoin(prevout);
```
Please dont delete comments. Just because a PR author decided against taking a suggestion doesnt mean review content should be deleted. Others may wish to agree with the suggestion or otherwise just view the history of what was discussed. Unless a comment is updated immediately afterwards, generally I'd prefer they not be edited.

danra commented:
utACK

TheBlueMatt commented:
re-utACK 4e955c58e13cfe089208f6b23b195d395ad99baa
